packages/liveblocks-core/src/__tests__/_utils.ts

```diff
@@ -141,58 +146,46 @@ export class MockWebSocketServer {
   /**
    * Create a new socket connection instance this server.
    */
-  public newSocket(callback?: (socket: MockWebSocket) => void): MockWebSocket {
-    const emitters = {
+  public newSocket(initFn?: (socket: MockWebSocket) => void): MockWebSocket {
+    const serverEvents = {
       onOpen: makeEventSource<IWebSocketEvent>(),
       onClose: makeEventSource<IWebSocketCloseEvent>(),
       onMessage: makeEventSource<IWebSocketMessageEvent>(),
       onError: makeEventSource<IWebSocketEvent>(),
     };
 
-    const publicListeners: Listeners = {
-      onOpen: emitters.onOpen.observable,
-      onClose: emitters.onClose.observable,
-      onMessage: emitters.onMessage.observable,
-      onError: emitters.onError.observable,
-    };
-
     const serverSocket: ServerSocket = {
       receivedMessages: this.receivedMessages,
-      accept: () => emitters.onOpen.notify(new Event("open")),
-      close: emitters.onClose.notify,
-      message: emitters.onMessage.notify,
-      error: emitters.onError.notify,
+      accept: () => serverEvents.onOpen.notify(new Event("open")),
+      close: serverEvents.onClose.notify,
+      send: serverEvents.onMessage.notify,
+      error: serverEvents.onError.notify,
     };
 
-    const socket = new MockWebSocket();
-    socket.linkToServerSocket(serverSocket, publicListeners);
-    this.connections.set(socket, emitters);
-    this.current = socket;
-
-    if (callback) {
-      this.onNewSocket(callback);
-    }
+    const clientSocket = new MockWebSocket();
+    clientSocket.linkToServerSocket(serverSocket, serverEvents);
+    this.connections.set(clientSocket, serverEvents);
+    this.current = clientSocket;
 
     // Run the callback in the next tick. This is important, because we first
     // need to return the socket.
     setTimeout(() => {
-      this.newSocketCallback?.(socket);
+      // Call the provided callback once, for this connection only.
+      initFn?.(clientSocket);
+
+      // ...then proceed to call the rest of the callbacks, which will be
+      // executed on every new connection
+      this.newConnectionCallbacks.notify({
```

Comment:
how does this get populated? It looks like the previous code wrapped the callback everytime it was set.

Reply:
Correct, and that was a bug. It gets populated by calling `wss.onConnection()`. In a unit test, you can use this to program how the server will behave once a connection gets made. For example:

![Screen Shot 2023-05-26 at 01 15 23@2x](https://github.com/liveblocks/liveblocks/assets/83844/588d1218-4b63-4a1f-9531-81ba84042348)

Reply:
The point of stubbing out this backend like in this example is to let us interact with and observe effects from the outside. Compare this rewritten test above to the same test as it currently exists on `main`, which fakes the behavior to test here by not using a socket connection at all, but instead making equivalent calls to room's internals.

https://github.com/liveblocks/liveblocks/blob/e0bd5a582370d2bec2b1e68e4873fa00a7332cd5/packages/liveblocks-core/src/__tests__/room.test.ts#L641-L661